Search: Use aliased Preact imports

jsnmoon opened the pull request:
#### Changes proposed in this Pull Request:
In preparation for using our Instant Search application within Gutenberg, we'll rely on using aliased Preact imports using our Webpack configuration. 

(In other words, our codebase will read as though it's importing React even though it's actually importing Preact.)

#### Jetpack product discussion
Necessary for our future customization efforts, see p1HpG7-c6N-p2.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Apply this change to your Jetpack site with a Search subscription.
* Ensure that the search application builds as expected.
* Ensure that the search interface works as expected.

Commit:
Use aliased Preact imports

Commit:
Changelog

bluefuton commented:
Looks good, but let's add some docs to help future us.

Can we add a 'React and Preact' section to the README explaining the background and how this is set up?

Commit:
Improve documentation

jsnmoon commented:
> Looks good, but let's add some docs to help future us.
> 
> Can we add a 'React and Preact' section to the README explaining the background and how this is set up? (@bluefuton)

Done!

bluefuton commented on projects/plugins/jetpack/modules/search/instant-search/README.md:
```diff
@@ -11,22 +11,28 @@ It enables the user to quickly jump into a full-page search experience with filt
+
+While the Instant Search application is primarily designed to be run as a Preact application, we use React imports in our code to ensure interoperability in React contexts, like in the Gutenberg editor.
+
+Using our Webpack configuration, we alias all `react` and `react-dom` to `preact/compat` to ensure that we ship Preact, not React, in our main bundle.
```
Excellent! 👍

bluefuton commented:
Oddly this seems to break for me in the Twenty Twenty One theme, with a seemingly unrelated error:

<img width="1501" alt="Screen Shot 2021-06-04 at 11 07 39" src="https://user-images.githubusercontent.com/17325/120723089-26c77600-c525-11eb-9cfe-59baf1991f80.png">

This doesn't happen on Jetpack `master`. Can you reproduce it too?

My build command was `pnpm build && pnpm build-search -- --watch`.

Commit:
Use direct preact imports for app injection code

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into update/search-app-import-convention

jsnmoon commented:
> Oddly this seems to break for me in the Twenty Twenty One theme, with a seemingly unrelated error:
> 
> <img alt="Screen Shot 2021-06-04 at 11 07 39" width="1501" src="https://user-images.githubusercontent.com/17325/120723089-26c77600-c525-11eb-9cfe-59baf1991f80.png">
> 
> This doesn't happen on Jetpack `master`. Can you reproduce it too?
> 
> My build command was `pnpm build && pnpm build-search -- --watch`. (@bluefuton)

I discovered that this was being caused by using `preact/compat`'s `{ render }` function in `index.jsx`, which behaves differently than `preact`'s `{ render }` function.

Since `index.jsx` is never used in a React context, I went ahead and reverted this change. Added the requisite code comment explaining the rationale as well.

jsnmoon commented:
Unfortunately, this breaks two of our existing Jest tests due to Jest using its own import resolutions. I'm currently working on a fix.

Commit:
Use React testing library

kangzj commented on projects/plugins/jetpack/tools/webpack.config.search.js:
```diff
@@ -105,6 +105,9 @@ module.exports = {
 			requestToExternal,
 			requestToHandle: defaultRequestToHandle,
 		} ),
+		new webpack.ProvidePlugin( {
```
Just curious where is `h` used?

jsnmoon replied:
Good catch, this code is actually no longer necessary. For context, `preact.h` corresponds to `react.createElement` (and stands for [hyperscript](https://jasonformat.com/wtf-is-jsx/)).

kangzj reviewed (dismissed):
It works well and looks good to me 👍

Commit:
Remove unnecessary ProvidePlugin instance

jeherve reviewed (approved):
This looks good to me. 👍

jeherve committed:
Merge remote-tracking branch 'origin/master' into update/search-app-import-convention

jsnmoon commented:
Thanks for merging, @sdixon194!

🚢 r227035-wpcom
